README.md
# CoSIAdata

### ExperimentHub package for CoSIA: **C**r**o**ss **S**pecies **I**nvestigation and **A**nalysis

[`CoSIA`](https://github.com/lasseignelab/CoSIA) is a package that provides researchers with an alternative 
methodology for comparing across species and tissues using normal wild-type RNA-Seq Gene Expression data from Bgee. 
Using RNA-Seq Gene Expression data, CoSIA provides multiple visualization tools to explore the transcriptome diversity and variation across genes, 
tissues, and species. CoSIA uses Coefficient of Variation and Shannon Entropy and Specificity to calculate transcriptome diversity and variation. 
CoSIA also provides additional conversion tools and utilities to provide a streamlined methodology for cross-species comparison 
across the tissues and genes of five commonly used biomedical research species 
(*Mus musculus*, *Rattus norvegicus*, *Danio rerio*, *Drosophila melanogaster*, and *Caenorhabditis elegans*) in addition to *Homo sapiens*.


## Getting Started

### Installing

**In preparation for the Bioconductor 3.17 release, we have developed CoSIA within a Bioconductor docker. 
Follow the instructions on [Bioconductor's Docker Help Page](https://www.bioconductor.org/help/docker/) to install and run CoSIAdata within the `bioconductor_docker:devel` container.**

In R:

``` r
## Install BiocManager if necessary; Install ExperimentHub
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ExperimentHub")

library(ExperimentHub)

## Store ExperimentHub in an object
eh <- ExperimentHub()
```

### Accessing CoSIAdata metadata

``` r
query(eh , "CoSIAdata")
```
Check out the vignette for more information on accesssing specific data in CoSIAdata.

## Authors

-   Anisha Haldar
-   Vishal H. Oza
-   Nathaniel DeVoss
-   Amanda D. Clark
-   Brittany N. Lasseigne

## Lasseigne Lab

[What is Happening in the Lasseigne Lab?](https://www.lasseigne.org/)

<img src="https://www.lasseigne.org/img/main/lablogo.png" width="75" height="75">

## Funding

This work was supported in part by the UAB Lasseigne Lab Start-Up funds (BNL, 
AH, ND, ADC and VHO), the UAB Pilot Center for Precision Animal Modeling (C-PAM)
(1U54OD030167) (BNL and VHO), UAB Pilot Center for Precision Animal Modeling 
(C-PAM) - Diversity Supplement (3U54OD030167-03S1) (ADC), and Mentored 
Experiences in Research, Instruction, and Teaching (MERIT) Program 
(K12 GM088010) (ADC).

## Acknowledgements

The authors would like to thank the members of the Lassigne lab for their 
support and feedback, in particular, Elizabeth J. Wilk, Jordan Whitlock, 
and Timothy C. Howton.

## License

This project is licensed under the MIT License - 
see the [LICENSE.md](LICENSE.md) file for details


NEWS.md
## Version 0.99.0

---

#New Package Release
CoSIAdata includes Variance Stabilized Transformation of Read Counts from 
Bgee RNA-Seq Expression Data across six species (*Homo sapiens*, *Mus musculus*,
*Rattus norvegicius*,*Danio rerio*, *Drosophila melanogaster*, and 
*Caenorhabditis elegans*) and more than 132 tissues. Each species has its own 
independent data frame with its unique set of tissue and gene specific 
expression data.
 
CoSIAdata is meant to be integrated into the CoSIA Package, a visualization 
tool for cross species comparison of expression metrics. However, it can be 
used to conduct independent species, tissue, and gene-specific 
expression analysis.

R/CoSIAdata-package.R
#' @keywords internal
#' @name CoSIAdata
#' @aliases CoSIAdata
#' @import ExperimentHub
#' @description 
#' CoSIAdata includes Variance Stabilized Transformation of Read Counts from 
#' Bgee RNA-Seq Expression Data across six species (Homo sapiens, Mus musculus,
#' Rattus norvegicius,Danio rerio, Drosophila melanogaster, and Caenorhabditis 
#' elegans) and more than 132 tissues. Each species has its own independent 
#' dataframe with its unique set of tissue and gene specific expression data.
#' 
#' CoSIAdata is meant to be integrated into the CoSIA Package, a visualization 
#' tool for cross species comparison of expression metrics. However, it can be 
#' used to conduct independent species, tissue, and gene-specific 
#' expression analysis.
#' 
#' Researchers might find this data useful as it provides Variance Stabilized 
#' Transformation of read counts to compare RNA-Seq expression profiles.
#' @import qpdf
#' @format Multiple dataframes
#' 
#' @examples
#' library(ExperimentHub)
#' eh = ExperimentHub()
#' query(eh, "CoSIAdata")
#' eh[["EH7863"]]
"_PACKAGE"

## usethis namespace: start
## usethis namespace: end
NULL


inst/scripts/make-data.R
# Experimental Hub Script for Downloading and Filtering RNA Seq Gene 
  # Expression data from Bgee

# install and load BgeeDB as well as other package that are needed to build this 
  # data frame (make sure that all packages are up to data)
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
BiocManager::install("BgeeDB")
library(BgeeDB)
install.packages("dplyr")
library(dplyr)
install.packages("magrittr")
library(magrittr)
install.packages("tidyr")
library(tidyr)
install.packages("tibble")
library(tibble)
install.packages("stringr")
library(stringr)

# Functions for Loading and Filtering Data
# This function downloads species specific data from Bgee, filters for 
  # necessary columns, and creates a column of the Variance 
  # Stabilized Transformations of Read Counts for this data.
Bgee_Download <- function(species, dataType) {
  bgee_db <- BgeeDB::Bgee$new(species = species, dataType = dataType)
  species_db <- BgeeDB::getData(bgee_db) # get the data as a bgee object
  # remove the quotations from the anatomical entities
  species_db$Anatomical.entity.name <- gsub('"', 
                                            "", 
                                            species_db$Anatomical.entity.name) 
  # filter the bgee data by these columns
  species_specific <- data.frame(dplyr::select(species_db, 
                                               Library.ID, Experiment.ID, 
                                               Gene.ID, Anatomical.entity.ID, 
                                               Anatomical.entity.name, TPM, 
                                               Read.count)) 
  VST_calc <- species_db %>% select(Gene.ID, 
                                    Library.ID, Anatomical.entity.name, 
                                    Read.count)
  VST_calc$Library.ID_Tissue <- paste(VST_calc$Library.ID, 
                                      VST_calc$Anatomical.entity.name, 
                                      sep = "_")
  VST_calc <- VST_calc %>% select(Gene.ID, Library.ID_Tissue, Read.count)
  VST_calc <- VST_calc %>%
    tidyr::pivot_wider(names_from = Library.ID_Tissue, 
                       values_from = Read.count) %>%
    column_to_rownames(var = "Gene.ID") %>%
    as.matrix() %>%
    round(., 0)
  VST_matrix <- as.data.frame(DESeq2::vst(VST_calc))
  VST_matrix <- tibble::rownames_to_column(VST_matrix, "Gene.ID")
  VST_df <- pivot_longer(VST_matrix, cols = -1, 
                         names_to = "Library.ID_Tissue", values_to = "VST")
  VST_df[c("Library.ID", "Anatomical.entity.name")] <- str_split_fixed(VST_df$Library.ID_Tissue, "_", 2)
  VST_df <- VST_df %>% select(Gene.ID, Library.ID, Anatomical.entity.name, VST)
  species_specific <- merge(VST_df, 
                            species_specific, 
                            by = c("Gene.ID", "Library.ID", 
                                   "Anatomical.entity.name"))
  return(species_specific)
}

# This function organizes the species specific data from above into 
  # the correct format for Experimental Hub.
Species_Dataframe <- function(dfBgee, species_name) {
  Species <- dfBgee %>%
    group_by(Gene.ID, Anatomical.entity.name) %>%
    summarise(Sample_size = n(), Median_VST = median(VST), 
              Scaled_Median_VST = median(scale(VST, center = FALSE)), 
              VST = paste(unique(VST), collapse = ", "), 
              Experiment_ID = paste(unique(Experiment.ID), collapse = ", "))
  Species_tissue <- dfBgee %>% select(Anatomical.entity.ID,
                                      Anatomical.entity.name)
  Species_tissue <- unique(Species_tissue)
  Species <- merge(Species, Species_tissue, by = "Anatomical.entity.name")
  names(Species)[names(Species) == "Anatomical.entity.name"] <- "Anatomical_entity_name"
  names(Species)[names(Species) == "Anatomical.entity.ID"] <- "Anatomical_entity_ID"
  names(Species)[names(Species) == "Gene.ID"] <- "Ensembl_ID"
  Species$Species <- species_name
  return(Species)
}

# Use functions to get Species-specific filtered BgeeDB

zebrafish_specific <- Bgee_Download("Danio_rerio", "rna_seq")
rat_specific <- Bgee_Download("Rattus_norvegicus", "rna_seq")
mouse_specific <- Bgee_Download("Mus_musculus", "rna_seq")
human_specific <- Bgee_Download("Homo_sapiens", "rna_seq")
fly_specific <- Bgee_Download("Drosophila_melanogaster", "rna_seq")
nematode_specific <- Bgee_Download("Caenorhabditis_elegans", "rna_seq")

# Get Species Experimental Hub Data
GEx_Bulk_Bgee_Zebrafish <- Species_Dataframe(zebrafish_specific, "Danio_rerio")
save(GEx_Bulk_Bgee_Zebrafish, file = "CoSIAdata/GEx_Bgee/EH_Dr.Rda")
GEx_Bulk_Bgee_Mouse <- Species_Dataframe(mouse_specific, "Mus_musculus")
save(GEx_Bulk_Bgee_Mouse, file = "CoSIAdata/GEx_Bgee/EH_Mm.Rda")
GEx_Bulk_Bgee_Rat <- Species_Dataframe(rat_specific, "Rattus_norvegicus")
save(GEx_Bulk_Bgee_Rat, file = "CoSIAdata/GEx_Bgee/EH_Rn.Rda")
GEx_Bulk_Bgee_Human <- Species_Dataframe(human_specific, "Homo_sapiens")
save(GEx_Bulk_Bgee_Human, file = "CoSIAdata/GEx_Bgee/EH_Hs.Rda")
GEx_Bulk_Bgee_Fly <- Species_Dataframe(fly_specific, "Drosophila_melanogaster")
save(GEx_Bulk_Bgee_Fly, file = "CoSIAdata/GEx_Bgee/EH_Dm.Rda")
GEx_Bulk_Bgee_Nematode <- Species_Dataframe(nematode_specific, 
                                            "Caenorhabditis_elegans")
save(GEx_Bulk_Bgee_Nematode, file = "CoSIAdata/GEx_Bgee/EH_Ce.Rda")


inst/scripts/make-metadata.R
### =========================================================================
### CoSIAdata metadata
### -------------------------------------------------------------------------
###

meta <- data.frame(
  Title = c(
    paste0(
      "VST normalized RNA-Sequencing data with annotations ",
      "for 362,533,125 Homo sapiens samples from Bgee"
    ),
    paste0(
      "VST normalized RNA-Sequencing data with annotations ",
      "for 31,405,642 Mus musculus samples from Bgee"
    ),
    paste0(
      "VST normalized RNA-Sequencing data with annotations ",
      "for 3,814,428 Rattus norvegicus samples from Bgee"
    ),
    paste0(
      "VST normalized RNA-Sequencing data with annotations ",
      "for 5,235,720 Danio rerio samples from Bgee"
    ),
    paste0(
      "VST normalized RNA-Sequencing data with annotations ",
      "for 4,576,399 Drosophila melanogaster samples from Bgee"
    ),
    paste0(
      "VST normalized RNA-Sequencing data with annotations ",
      "for 1,923,064 Caenorhabditis elegans samples from Bgee"
    )
  ),
  Description = c(
    paste0(
      "RNA-seq Variance Stabilized Transformation of Read ",
      "Counts Expression Data and Annotations for 362,533,125 ",
      "Homo sapiens samples across 132 tissues, represented as ",
      "an RData file. R data representation derieved from Bgee."
    ),
    paste0(
      "RNA-seq Variance Stabilized Transformation of Read ",
      "Counts Expression Data and Annotations for 31,405,642 ",
      "Mus musculus samples across 71 tissues, represented as ",
      "an RData file. R data representation derieved from Bgee."
    ),
    paste0(
      "RNA-seq Variance Stabilized Transformation of Read ",
      "Counts Expression Data and Annotations for 3,814,428 ",
      "Rattus norvegicus samples across 21 tissues, represented ",
      "as an RData file. R data representation derieved from Bgee."
    ),
    paste0(
      "RNA-seq Variance Stabilized Transformation of Read ",
      "Counts Expression Data and Annotations for 5,235,720 ",
      "Danio rerio samples across 25 tissues, represented as ",
      "an RData file. R data representation derieved from Bgee."
    ),
    paste0(
      "RNA-seq Variance Stabilized Transformation of Read ",
      "Counts Expression Data and Annotations for 4,576,399 ",
      "Drosophila melanogaster samples across 18 tissues, represented",
      "as an RData file. R data representation derieved from Bgee."
    ),
    paste0(
      "RNA-seq Variance Stabilized Transformation of Read Counts",
      "Expression Data and Annotations for 1,923,064",
      "Caenorhabditis elegans samples across 5 tissues represented as",
      "an RData file. R data representation derieved from Bgee."
    )
  ),
  BiocVersion = rep("3.17", 6),
  Genome = c("GRCh38.p13", "GRCm38.p6", "Rnor_6.0", "GRCz11", 
             "BDGP6.28", "WBcel235"),
  SourceType = rep("tar.gz", 6),
  SourceUrl = c(
    "https://bgee.org/ftp/bgee_v15_0/download/processed_expr_values/rna_seq/Homo_sapiens/Homo_sapiens_RNA-Seq_read_counts_TPM_FPKM.tar.gz",
    "https://bgee.org/ftp/bgee_v15_0/download/processed_expr_values/rna_seq/Mus_musculus/Mus_musculus_RNA-Seq_read_counts_TPM_FPKM.tar.gz",
    "https://bgee.org/ftp/bgee_v15_0/download/processed_expr_values/rna_seq/Rattus_norvegicus/Rattus_norvegicus_RNA-Seq_read_counts_TPM_FPKM.tar.gz",
    "https://bgee.org/ftp/bgee_v15_0/download/processed_expr_values/rna_seq/Danio_rerio/Danio_rerio_RNA-Seq_read_counts_TPM_FPKM.tar.gz",
    "https://bgee.org/ftp/bgee_v15_0/download/processed_expr_values/rna_seq/Drosophila_melanogaster/Drosophila_melanogaster_RNA-Seq_read_counts_TPM_FPKM.tar.gz",
    "https://bgee.org/ftp/bgee_v15_0/download/processed_expr_values/rna_seq/Caenorhabditis_elegans/Caenorhabditis_elegans_RNA-Seq_read_counts_TPM_FPKM.tar.gz"
  ),
  SourceVersion = rep("Bgee v15.0", 6),
  Species = c("Homo sapiens", "Mus musculus", "Rattus norvegicus", 
              "Danio rerio", "Drosophila melanogaster", 
              "Caenorhabditis elegans"),
  TaxonomyId = c(9606, 10090, 10116, 7955, 7227, 6239),
  Coordinate_1_based = FALSE,
  DataProvider = "Bgee",
  Maintainer = "Amanda D. Clark <amanda@freshairfamily.org>",
  RDataClass = c(rep("data.frame", 6)),
  DispatchClass = c(rep("Rda", 6)),
  RDataPath = c(
    "CoSIAdata/GEx_Bgee/EH_Hs.RData",
    "CoSIAdata/GEx_Bgee/EH_Mm.RData",
    "CoSIAdata/GEx_Bgee/EH_Rn.RData",
    "CoSIAdata/GEx_Bgee/EH_Dr.RData",
    "CoSIAdata/GEx_Bgee/EH_Dm.RData",
    "CoSIAdata/GEx_Bgee/EH_Ce.RData"
  )
)

write.csv(meta, file = "inst/extdata/metadata.csv", row.names = FALSE)
